src/Components/Web.JS/src/Rendering/DomMerging/DomSync.ts

```diff
@@ -284,11 +290,21 @@ function domNodeComparer(a: Node, b: Node): UpdateCost {
       // For elements, we're only doing a shallow comparison and don't know if attributes/descendants are different.
       // We never 'update' one element type into another. We regard the update cost for same-type elements as zero because
       // then the 'find common prefix/suffix' optimization can include elements in those prefixes/suffixes.
+      // If an element has the 'data-permanent' attribute, it may have content that should be retained between DOM synchronizations.
+      // To maximize the chance that data permanent elements get matched up correctly, we compare the IDs of each element.
       // TODO: If we want to support some way to force matching/nonmatching based on @key, we can add logic here
       //       to return UpdateCost.Infinite if either has a key but they don't match. This will prevent unwanted retention.
       //       For the converse (forcing retention, even if that means reordering), we could post-process the list of
       //       inserts/deletes to find matches based on key to treat those pairs as 'move' operations.
-      return (a as Element).tagName === (b as Element).tagName ? UpdateCost.None : UpdateCost.Infinite;
+      if ((a as Element).tagName !== (b as Element).tagName) {
+        return UpdateCost.Infinite;
+      }
+
+      if (areIncompatibleDataPermanentElements(a as Element, b as Element)) {
+        return UpdateCost.Infinite;
+      }
+
+      return UpdateCost.None;
```

Comment:
As well as having the E2E tests, I think we should have Jest unit tests for the `data-permanent` behaviors. Would be a lot easier to cover edge cases or quickly detect if this specific thing regresses in the future.